### DISCUSSION TITLE
Emulating a MicroPython device in order to utilize Thonny's file explorer features

### DISCUSSION THREAD

**Discussion by liudr** [thonny/thonny#3418] [General]
I admit that my topic is a bit strange so please allow me to explain. Here is what I am facing: I have been developing MicroPython programs for a number of years and have been using Thonny primarily for tests and file managements on the MP boards.

So on a typical day I would load mp and config files I've just updated to my MP board via thonny's file manager and reboot my board with machine.reset() etc. and watch the printouts on thonny, interrupting it sometimes to check out things. I feel that thonny's file manager is indispensable because I either update mp files or config files with it.

I have designed a number of mp boards each featuring something, such as a relay board, a sensor board etc.

So now I need to develop a board (ESP32) that acts as an IR remote control which doesn't have micropython support. The support is purely Arduino IDE with C++ and ESP32 for Arduino. This development is ongoing but in order to give this board the same look and feel as the other boards running mp, I'd like to be able to save and edit a config file on the board via thonny.

I know thonny only works with micropython boards by sending micropython code to the board to execute and receive results from it. But I've probed the communications between thonny and an mp board, and read the raw paste REPL mode doc (very bare minimal provided). My idea is to emulate the responses of the mp raw paste repl mode so that thonny will get what it asks in order to provide the file manager features. This will involve at least capture the initial thonny helper class code and understand subsequent lines of code that does getcwd, read and write files, etc. There's also the list of builtin modules etc. I wonder if anyone has done something like this in the past, an mp emulator for the sake of file transfers :D

**Comment by aivarannamaa** [comment]
Why not use full MP instead of emulator on this ESP?

**Reply by liudr** [reply]
Most of my boards are running MP. This new board is running esp32 for Arduino which is c++, not MP. The new work I am doing doesn't have MP support, only esp32 for Arduino. Unless I can run two apps together on an esp32, I can't use MP if I must use esp32 for Arduino.

**Comment by aivarannamaa** [comment]
This problem doesn't look specific to Thonny, so you might get better answers at https://github.com/orgs/micropython/discussions
